apps/aecore/lib/aecore/sync/chain.ex

```diff
@@ -0,0 +1,130 @@
+defmodule Aecore.Sync.Chain do
+  @moduledoc """
+  Implements all the functions regarding the Chain structure of the SyncTask
+  """
+
+  alias Aecore.Chain.Header
+  alias Aecore.Chain.BlockValidation
+  alias Aecore.Sync.Task
+  alias Aeutil.List, as: ListUtils
+  alias __MODULE__
+
+  @type peer_id :: pid()
+  @type chain_id :: reference()
+  @type height :: non_neg_integer()
+  @type header_hash :: binary()
+
+  @typedoc "Holds data for header height and hash"
+  @type chain :: %{height: height(), hash: header_hash()}
+
+  @type t :: %Chain{chain_id: chain_id(), peers: list(peer_id()), chain: list(chain())}
+
+  defstruct chain_id: nil,
+            peers: [],
+            chain: []
+
+  @spec init_chain(peer_id(), Header.t()) :: Chain.t()
+  def init_chain(peer_id, header) do
+    init_chain(Kernel.make_ref(), [peer_id], header)
+  end
+
+  @spec init_chain(chain_id(), peer_id(), Header.t()) :: Chain.t()
+  def init_chain(chain_id, peers, %Header{height: height, prev_hash: prev_hash} = header) do
+    header_hash = BlockValidation.block_header_hash(header)
+
+    prev_header_data =
+      if height > 1 do
+        [%{height: height - 1, hash: prev_hash}]
+      else
+        []
+      end
+
+    %Chain{
+      chain_id: chain_id,
+      peers: peers,
+      chain: [%{height: height, hash: header_hash}] ++ prev_header_data
+    }
+  end
+
+  @spec merge_chains(Chain.t(), Chain.t()) :: Chain.t()
+  def merge_chains(%Chain{chain_id: chain_id, peers: peers_1, chain: chain_1}, %Chain{
+        chain_id: chain_id,
+        peers: peers_2,
+        chain: chain_2
+      }) do
+    peers =
+      (peers_1 ++ peers_2)
+      |> Enum.sort()
+      |> Enum.uniq()
```

Comment:
The peers are already sorted -> We can use this to merge and deduplicate the peers in O(n) instead of O(n log n)

Reply:
If the two variables of peers_1 and peers_2 are sorted when concatenating them they won't be anymore: `[2, 3] ++ [1, 4] = [2, 3, 1, 4]`. So I don't really understand your point here.

Reply:
No, you don't need to concatenate the list - all you need to do is to perform the merging step from the classic merge sort algorithm and add some more checks to make sure that while merging you aren't including duplicates. Here is my code(with tests) for performing this operation in linear time: https://gist.github.com/grzegorz225/431bf22243f032843d13c0ab03a2396d
On my computer it runs 3-4 times faster for random lists containing 10k elements than the naive algorithm.

Reply:
Okay, but as far as I know, we are not going to have more than 30 peers connected to us. So this operation will actually happen for 30 elements. I don't think that it makes much sense to change it for such a little speed change.

Reply:
I don't know if there is much reason in sorting them tbo. But have in mind that most of the times this elements will not be more than 10-20. We can only connect to 30 peers AFAIK.